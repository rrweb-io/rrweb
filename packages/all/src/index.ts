export * from 'rrweb';
export * from '@rrweb/packer';
// export * from '@rrweb/rrweb-plugin-console-record';
// export * from '@rrweb/rrweb-plugin-console-replay';
