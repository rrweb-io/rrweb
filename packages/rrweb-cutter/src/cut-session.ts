import { NodeType } from 'rrweb-snapshot';
import {
  adoptedStyleSheetData,
  eventWithTime,
  MediaInteractions,
  styleDeclarationData,
  styleSheetRuleData,
} from '@rrweb/types';
import {
  RRCanvasElement,
  RRStyleElement,
  RRMediaElement,
  RRIFrameElement,
  RRDocument,
} from 'rrdom';
import type { IRRNode, RRElement } from 'rrdom';
import { IncrementalSource, EventType, SyncReplayer } from 'rrweb';
import { playerConfig } from 'rrweb/typings/types';
import cloneDeep from 'lodash.clonedeep';
import snapshot from './snapshot';
export type CutterConfig = {
  points: number[];
  // config for the Sync Replayer
  replayerConfig?: Partial<playerConfig>;
  /**
   * If true, the sequentialId of the events will be updated.
   * If a string, the sequentialId will be updated with the string as the key.
   */
  updateSequentialId?: boolean | string;
  onSessionCut?: (context: {
    replayer: SyncReplayer;
    cutSession: SessionCut;
    originalEvents: eventWithTime[];
  }) => SessionCut;
  customEventsHandler?: (context: {
    replayer: SyncReplayer;
    events: eventWithTime[];
    currentTimestamp: number;
    cutTimestamp: number;
  }) => eventWithTime[];
};

export type SessionCut = {
  events: eventWithTime[];
  startTimestamp: number;
  endTimestamp: number;
  startTime: number;
  endTime: number;
};

export function cutSession(
  events: eventWithTime[],
  config: CutterConfig,
): SessionCut[] {
  const baseTimestamp = events[0]?.timestamp || 0;
  const defaultResult = [
    wrapCutSession(
      events,
      config,
      baseTimestamp,
      events[events.length - 1]?.timestamp || 0,
      baseTimestamp,
    ),
  ];
  // Events length is too short so that cutting process is not needed.
  if (events.length < 2) return defaultResult;

  const { points } = config;
  if (!points || points.length == 0) return defaultResult;

  events = events.sort((a1, a2) => a1.timestamp - a2.timestamp);
  const totalTime = events[events.length - 1].timestamp - events[0].timestamp;

  const validSortedPoints = getValidSortedPoints(points, totalTime);
  if (validSortedPoints.length < 1) return defaultResult;
  const validSortedTimestamp = validSortedPoints.map(
    (point) => baseTimestamp + point,
  );

  const results: SessionCut[] = [];
  const replayer = new SyncReplayer(events, {
    ...config.replayerConfig,
  });
  let cutPointIndex = 0;
  replayer.play(({ index }) => {
    if (index + 1 >= events.length) return false;

    const nextEvent = events[index + 1];

    let currentTimestamp = validSortedTimestamp[cutPointIndex];
    let eventsCache: eventWithTime[] | null = null;
    /**
     * This loop is for the situation that cutting points are in the middle gap between two events.
     * These cut points have the same cut events so there is no need to generate them for each cut point.
     * We can cache the them and use them for the next cut point.
     */
    while (
      cutPointIndex < validSortedTimestamp.length &&
      // If the next event exceed the cut point, we need to cut the events with the current replayer status.
      nextEvent.timestamp > validSortedTimestamp[cutPointIndex]
    ) {
      if (results.length === 0) {
        results.push(
          wrapCutSession(
            events
              .slice(0, index + 1)
              .filter((e) => e.timestamp < validSortedTimestamp[cutPointIndex]),
            config,
            baseTimestamp,
            currentTimestamp,
            baseTimestamp,
          ),
        );
      }
      cutPointIndex++;
      const nextCutTimestamp =
        cutPointIndex < validSortedTimestamp.length
          ? validSortedTimestamp[cutPointIndex]
          : events[events.length - 1].timestamp;
      if (eventsCache !== null) {
        const timeDiff = currentTimestamp - eventsCache[0].timestamp;
        const newEvents = eventsCache.map((e) => ({
          ...e,
          timestamp:
            e.timestamp < currentTimestamp
              ? e.timestamp + timeDiff
              : e.timestamp,
        }));
        results.push(
          wrapCutSession(
            newEvents,
            config,
            currentTimestamp,
            nextCutTimestamp,
            baseTimestamp,
          ),
        );
        continue;
      }

      const session = cutEvents(
        events.slice(index + 1).filter((e) => {
          // If this would be the last session, include all of the rest events.
          if (cutPointIndex >= validSortedTimestamp.length) return true;
          return e.timestamp < nextCutTimestamp;
        }),
        replayer,
        config,
        currentTimestamp,
        nextCutTimestamp,
        baseTimestamp,
      );
      results.push(session);
      eventsCache = session.events;
      currentTimestamp = nextCutTimestamp;
    }
    return cutPointIndex < validSortedTimestamp.length;
  });
  return results;
}

export function getValidSortedPoints(points: number[], totalTime: number) {
  const validSortedPoints = [];
  for (let i = 0; i < points.length; i++) {
    const point = points[i];
    if (point <= 0 || point >= totalTime) continue;
    validSortedPoints.push(point);
  }
  // The default sortting function will sort the array as string.
  return validSortedPoints.sort((a, b) => a - b);
}

function wrapCutSession(
  events: eventWithTime[],
  config: CutterConfig,
  startTimestamp: number,
  endTimestamp: number,
  baseTimestamp: number,
): SessionCut {
  let clonedEvents = cloneDeep(events);
  if (config.updateSequentialId) {
    const key =
      typeof config.updateSequentialId === 'string'
        ? config.updateSequentialId
        : '_sid';
    let sequentialId = 0;
    clonedEvents = events.map((e) => {
      Object.assign(e, {
        [key]: ++sequentialId,
      });
      return e;
    });
  }
  return {
    events: clonedEvents,
    startTimestamp,
    endTimestamp,
    startTime: startTimestamp - baseTimestamp,
    endTime: endTimestamp - baseTimestamp,
  };
}

/**
 * Cut original events at the cutting point which will produce two parts.
 * Only return the deep cloned cut session before the cutting point (the previous part).
 * @param events - The events to be cut.
 * @param replayer - The sync replayer instance.
 * @param currentTimestamp - The current timestamp.
 * @param cutTimestamp - The timestamp to cut.
 */
function cutEvents(
  events: eventWithTime[],
  replayer: SyncReplayer,
  config: CutterConfig,
  currentTimestamp: number,
  cutTimestamp: number,
  baseTimestamp: number,
) {
  let result: eventWithTime[] = [];
  if (replayer.latestMetaEvent) {
    const metaEvent = { ...replayer.latestMetaEvent };
    metaEvent.timestamp = currentTimestamp;
    result.push(metaEvent);
  }
  const FullsnapshotDelay = 1, // The delay between MetaEvent and FullSnapshot.
    // The delay between FullSnapshot and MetaEvent. Make sure the iframe event is after the full snapshot.
    IFrameEventDelay = 2,
    // The delay between IncrementalSnapshot and MetaEvent. Make sure the incremental event is applied after all frames are loaded.
    IncrementalEventDelay = 3;

  const iframeSnapshots: eventWithTime[] = [];
  const incrementalEvents: eventWithTime[] = [];
  const onSerialize = (n: IRRNode) => {
    const timestamp = currentTimestamp + IncrementalEventDelay;
    if (n.RRNodeType === NodeType.Document) {
      const rrDoc = n as RRDocument;
      rrDoc.scrollData &&
        incrementalEvents.push({
          type: EventType.IncrementalSnapshot,
          data: rrDoc.scrollData,
          timestamp,
        });
      return;
    }
    const rrElement = n as RRElement;
    rrElement.inputData &&
      incrementalEvents.push({
        type: EventType.IncrementalSnapshot,
        data: rrElement.inputData,
        timestamp,
      });
    rrElement.scrollData &&
      incrementalEvents.push({
        type: EventType.IncrementalSnapshot,
        data: rrElement.scrollData,
        timestamp,
      });
    if (rrElement instanceof RRCanvasElement)
      rrElement.canvasMutations.forEach((canvasData) =>
        incrementalEvents.push({
          type: EventType.IncrementalSnapshot,
          data: canvasData.mutation,
          timestamp,
        }),
      );
    else if (rrElement instanceof RRStyleElement)
      rrElement.rules.forEach((styleRule) =>
        incrementalEvents.push({
          type: EventType.IncrementalSnapshot,
          data: styleRule,
          timestamp,
        }),
      );
    else if (rrElement instanceof RRMediaElement) {
      (rrElement.volume !== undefined || rrElement.muted !== undefined) &&
        incrementalEvents.push({
          type: EventType.IncrementalSnapshot,
          data: {
            source: IncrementalSource.MediaInteraction,
            type: MediaInteractions.VolumeChange,
            volume: rrElement.volume,
            muted: rrElement.muted,
            id: replayer.getMirror().getId(n),
          },
          timestamp,
        });
      rrElement.playbackRate !== undefined &&
        incrementalEvents.push({
          type: EventType.IncrementalSnapshot,
          data: {
            source: IncrementalSource.MediaInteraction,
            type: MediaInteractions.RateChange,
            playbackRate: rrElement.playbackRate,
            id: replayer.getMirror().getId(n),
          },
          timestamp,
        });
    } else if (rrElement instanceof RRIFrameElement) {
      if (!rrElement.contentDocument) return;
      const iframeSnapshot = snapshot(rrElement.contentDocument, {
        mirror: replayer.getMirror(),
        onSerialize,
      });
      if (!iframeSnapshot) return;
      iframeSnapshots.push({
        type: EventType.IncrementalSnapshot,
        data: {
          source: IncrementalSource.Mutation,
          adds: [
            {
              parentId: replayer.getMirror().getId(rrElement),
              nextId: null,
              node: iframeSnapshot,
            },
          ],
          removes: [],
          texts: [],
          attributes: [],
          isAttachIframe: true,
        },
        timestamp: currentTimestamp + IFrameEventDelay,
      });
    }
  };

  let fullSnapshot = snapshot(replayer.virtualDom, {
    mirror: replayer.getMirror(),
    onSerialize,
  });
  if (!fullSnapshot) {
    console.warn(
      `Failed to generate full snapshot at timestamp ${currentTimestamp}. Using a blank snapshot as fallback.`,
    );
    // Fallback to a blank snapshot.
    fullSnapshot = {
      type: NodeType.Document,
      childNodes: [],
      id: 1,
    };
  }
  const fullSnapshotEvent: eventWithTime = {
    type: EventType.FullSnapshot,
    data: {
      node: fullSnapshot,
      initialOffset: {
        top: replayer.virtualDom.scrollTop,
        left: replayer.virtualDom.scrollLeft,
      },
    },
    timestamp: currentTimestamp + FullsnapshotDelay,
  };
  result.push(fullSnapshotEvent);
  result = result.concat(iframeSnapshots, incrementalEvents);
  result = result.concat(
    replayer.unhandledEvents.map((e) => ({
      ...e,
      timestamp: currentTimestamp + IncrementalEventDelay,
    })),
  );
  result = result.concat(
    filterAdoptedStyleData(
      replayer,
      replayer.adoptedStyleSheets,
      replayer.constructedStyleMutations,
      events
        .filter(
          (event) =>
            event.timestamp <= cutTimestamp &&
            event.type === EventType.IncrementalSnapshot &&
            event.data.source === IncrementalSource.AdoptedStyleSheet,
        )
        .map((event) => event.data as adoptedStyleSheetData),
      currentTimestamp + IncrementalEventDelay,
    ),
  );

  replayer.mousePos &&
    result.push({
      type: EventType.IncrementalSnapshot,
      data: {
        source: IncrementalSource.MouseMove,
        positions: [
          {
            x: replayer.mousePos.x,
            y: replayer.mousePos.y,
            id: replayer.mousePos.id,
            timeOffset: 0,
          },
        ],
      },
      timestamp: currentTimestamp + IncrementalEventDelay,
    });

  replayer.lastSelectionData &&
    result.push({
      type: EventType.IncrementalSnapshot,
      data: replayer.lastSelectionData,
      timestamp: currentTimestamp + IncrementalEventDelay,
    });
  if (config.customEventsHandler)
    try {
      result = result.concat(
        config.customEventsHandler({
          events,
          replayer,
          currentTimestamp,
          cutTimestamp,
        }),
      );
    } catch (e) {
      warn(config, 'Custom Event Handler error: ', e);
    }
  result = result.concat(events);
  let session = wrapCutSession(
    result,
    config,
    currentTimestamp,
    cutTimestamp,
    baseTimestamp,
  );
  if (config.onSessionCut)
    try {
      session = config.onSessionCut({
        cutSession: session,
        replayer,
        originalEvents: events,
      });
    } catch (e) {
      warn(config, 'Custom Event Handler error: ', e);
    }
  return session;
}

/**
 * Keep the adopted style data and style mutations which are still valid at this time point.
 */
function filterAdoptedStyleData(
  replayer: SyncReplayer,
  adoptedStyleSheetData: adoptedStyleSheetData[],
  constructedStyleMutations: (styleSheetRuleData | styleDeclarationData)[],
  upcomingAdoptedStyleSheets: adoptedStyleSheetData[],
  timestamp: number,
) {
  const events: eventWithTime[] = [];
  // A map from style id without style construction to the owner node id.
  const noConstructStyleIdToNodeIdMap = new Map<number, number>();
  const builtStyleIds = new Set<number>();
  const mirror = replayer.getMirror();
  /**
   * Reversely iterate the adopted style sheet data to find valid styles.
   * @param adoptedStyleSheetData - The adopted style sheet data.
   * @param onValidData - The callback function when the data is valid.
   */
  const reverseIterateAdoptedStyleData = (
    adoptedStyleSheetData: adoptedStyleSheetData[],
    onValidData?: (data: adoptedStyleSheetData) => void,
  ) => {
    for (let i = adoptedStyleSheetData.length - 1; i >= 0; i--) {
      const adoptedStyleSheet = adoptedStyleSheetData[i];
      if (mirror.has(adoptedStyleSheet.id)) {
        adoptedStyleSheet.styleIds?.forEach((styleId: number) => {
          if (builtStyleIds.has(styleId)) return;
          noConstructStyleIdToNodeIdMap.set(styleId, adoptedStyleSheet.id);
        });
        adoptedStyleSheet.styles?.forEach((styleConstruction) => {
          if (noConstructStyleIdToNodeIdMap.has(styleConstruction.styleId))
            noConstructStyleIdToNodeIdMap.delete(styleConstruction.styleId);
        });
        onValidData?.(adoptedStyleSheet);
      }
      // The owner node of style data has been removed but its style construction may still be used by other adoptedStyleSheets.
      else {
        if (!adoptedStyleSheet.styles) continue;
        adoptedStyleSheet.styles.forEach((style) => {
          if (builtStyleIds.has(style.styleId)) return;
          const ownerNodeId = noConstructStyleIdToNodeIdMap.get(style.styleId);
          if (!ownerNodeId) return;
          noConstructStyleIdToNodeIdMap.delete(style.styleId);
          builtStyleIds.add(style.styleId);
          onValidData?.({
            source: IncrementalSource.AdoptedStyleSheet,
            id: ownerNodeId,
            styles: [style],
            styleIds: [style.styleId],
          });
        });
      }
    }
  };
  reverseIterateAdoptedStyleData(upcomingAdoptedStyleSheets);
  reverseIterateAdoptedStyleData(adoptedStyleSheetData, (data) => {
    events.unshift({
      type: EventType.IncrementalSnapshot,
      data,
      timestamp,
    });
  });
  const styleMutationDelay = 1;
  constructedStyleMutations.forEach((styleConstruction) => {
    if (
      styleConstruction.styleId &&
      builtStyleIds.has(styleConstruction.styleId)
    )
      events.push({
        type: EventType.IncrementalSnapshot,
        data: styleConstruction,
        timestamp: timestamp + styleMutationDelay,
      });
  });
  return events;
}

function warn(config: CutterConfig, ...args: unknown[]) {
  const logger = config.replayerConfig?.logger || console;
  if (config.replayerConfig?.showWarning || config.replayerConfig?.showDebug)
    logger.warn(args);
}
