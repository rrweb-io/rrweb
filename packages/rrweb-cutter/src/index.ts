export * from './cut-session';
export * from './prune-branches';
