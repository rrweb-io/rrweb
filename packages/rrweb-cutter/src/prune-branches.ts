import { serializedNodeWithId } from 'rrweb-snapshot';
import type {
  addedNodeMutation,
  eventWithTime,
  mousePosition,
} from '@rrweb/types';
import { IncrementalSource } from 'rrweb';
import { EventType } from 'rrweb';
import { SyncReplayer } from 'rrweb';

export function pruneBranches(
  events: eventWithTime[],
  { keep }: { keep: number[] },
): eventWithTime[] {
  const result: eventWithTime[] = [];
  const replayer = new SyncReplayer(events);
  const treeSet = new Set<number>(keep);
  replayer.reversePlay(({ event }) => {
    if (event.type === EventType.FullSnapshot) {
      const { node } = event.data;
      const tree = getTreeForId(treeSet, node, keep);
      tree.forEach((id) => treeSet.add(id));
    } else if (event.type === EventType.IncrementalSnapshot) {
      if (event.data.source === IncrementalSource.Mutation) {
        const { adds, removes } = event.data;
        removes.forEach((remove) => {
          if (treeSet.has(remove.id)) treeSet.add(remove.parentId);
        });
        adds.forEach((add) => {
          const tree = getTreeForId(treeSet, add.node, keep);
          if (tree.size) {
            treeSet.add(add.parentId);
            tree.forEach((id) => treeSet.add(id));
          } else if (
            'childNodes' in add.node &&
            add.node.childNodes.length > 0
          ) {
            const tree = getTreeForId(treeSet, add.node, keep);
            if (tree.size) treeSet.add(add.parentId);
            tree.forEach((id) => treeSet.add(id));
          }
        });
      }
    }
    return true;
  });

  replayer.play(({ event }) => {
    if (
      [EventType.Meta, EventType.Load, EventType.DomContentLoaded].includes(
        event.type,
      )
    ) {
      result.push(event);
    } else if (event.type === EventType.FullSnapshot) {
      const { node } = event.data;
      const prunedNode = reconstructTreeWithIds(node, treeSet);
      if (prunedNode)
        result.push({
          ...event,
          data: {
            ...event.data,
            node: prunedNode,
          },
        } as eventWithTime);
    } else if (event.type === EventType.IncrementalSnapshot) {
      if ('positions' in event.data) {
        const { positions } = event.data;
        const prunedPositions: mousePosition[] = positions.filter((p) =>
          treeSet.has(p.id),
        );
        if (prunedPositions.length > 0)
          result.push({
            ...event,
            data: {
              ...event.data,
              positions: prunedPositions,
            },
          } as eventWithTime);
      } else if ('id' in event.data) {
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        if (treeSet.has(event.data.id!)) result.push(event);
      } else if (event.data.source === IncrementalSource.Mutation) {
        const { removes, adds, texts, attributes } = event.data;
        const prunedRemoves = removes.filter((remove) =>
          treeSet.has(remove.id),
        );
        const prunedAdds = adds
          .map((add) =>
            treeSet.has(add.parentId) && keep.includes(add.parentId)
              ? add
              : {
                  ...add,
                  node: reconstructTreeWithIds(add.node, treeSet),
                },
          )
          .filter((add) => Boolean(add.node)) as addedNodeMutation[];
        const prunedTexts = texts.filter((text) => treeSet.has(text.id));
        const prunedAttributes = attributes.filter((attr) =>
          treeSet.has(attr.id),
        );
        if (
          prunedRemoves.length > 0 ||
          prunedAdds.length > 0 ||
          prunedTexts.length > 0 ||
          prunedAttributes.length > 0
        )
          result.push({
            ...event,
            data: {
              ...event.data,
              removes: prunedRemoves,
              adds: prunedAdds,
              texts: prunedTexts,
              attributes: prunedAttributes,
            },
          } as eventWithTime);
      }
    }
    return true;
  });
  return result;
}

export function getTreeForId(
  treeSet: Set<number>,
  node: serializedNodeWithId,
  keepIds: number[],
): Set<number> {
  const results = new Set<number>();
  if (treeSet.has(node.id)) {
    getIdsInNode(node, keepIds).forEach((id) => results.add(id));
  } else if ('childNodes' in node) {
    for (let i = 0; i < node.childNodes.length; i++) {
      const child = node.childNodes[i];
      const childTree = getTreeForId(treeSet, child, keepIds);
      if (childTree.size > 0) {
        results.add(node.id);
        childTree.forEach((id) => results.add(id));
      }
    }
  }
  return results;
}

export function getIdsInNode(
  node: serializedNodeWithId,
  keepIds: number[],
): Array<number> {
  const results: number[] = [];
  results.push(node.id);
  if (keepIds.includes(node.id) && 'childNodes' in node) {
    for (let i = 0; i < node.childNodes.length; i++) {
      const child = node.childNodes[i];
      results.push(...getIdsInNode(child, keepIds));
    }
  }
  return results;
}

export function reconstructTreeWithIds(
  node: serializedNodeWithId,
  ids: Set<number>,
): serializedNodeWithId | undefined {
  if (ids.has(node.id)) {
    if ('childNodes' in node) {
      node.childNodes = node.childNodes
        .map((child) => reconstructTreeWithIds(child, ids))
        .filter(Boolean) as serializedNodeWithId[];
    }
    return node;
  }
  return undefined;
}
