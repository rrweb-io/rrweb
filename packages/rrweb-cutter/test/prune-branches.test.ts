/**
 * @jest-environment jsdom
 */
import { NodeType } from 'rrweb-snapshot';
import { EventType } from 'rrweb';
import { SyncReplayer } from 'rrweb';
import type { eventWithTime } from '@rrweb/types';
import { printRRDom } from 'rrdom';
import { eventsFn as inlineStyleEvents } from './events/inline-style.event';
import { pruneBranches } from '../src';
import { assertSnapshot } from 'rrweb/test/utils';

describe('prune branches', () => {
  it("should cut branches that doesn't include id", () => {
    const events = inlineStyleEvents() as eventWithTime[];
    const result = pruneBranches(events, { keep: [14] });
    expect(result).toHaveLength(5);
    const replayer = new SyncReplayer(result);
    replayer.play();

    expect(
      printRRDom(replayer.virtualDom, replayer.getMirror()),
    ).toMatchSnapshot('pruned all but 14');
  });

  it("should remove mutations that don't include ids", () => {
    const mutationEvent = {
      type: EventType.IncrementalSnapshot,
      data: {
        source: 0,
        texts: [
          {
            id: 15,
            value: 'Kept',
          },
          {
            id: 1001,
            value: 'Cut',
          },
        ],
        attributes: [
          {
            id: 14,
            attributes: {
              'data-attr': 'Kept',
            },
          },
          {
            id: 1002,
            attributes: {
              'data-attr': 'Cut',
            },
          },
        ],
        removes: [
          {
            parentId: 14,
            id: 15,
          },
          {
            parentId: 1002,
            id: 1003,
          },
        ],
        adds: [
          {
            parentId: 14,
            nextId: null,
            node: {},
          },
          {
            parentId: 1001,
            nextId: null,
            node: {},
          },
        ],
      },
    };
    const events = [...inlineStyleEvents(), mutationEvent] as eventWithTime[];
    const result = pruneBranches(events, { keep: [14] });
    expect(result[result.length - 1]).toMatchSnapshot();
  });

  it('should remove branches based on nodes that came in after fullsnapshot', () => {
    const mutationEvent = {
      type: EventType.IncrementalSnapshot,
      data: {
        source: 0,
        texts: [],
        attributes: [],
        removes: [],
        adds: [
          {
            parentId: 14,
            nextId: null,
            node: {
              id: 99,
              type: NodeType.Element,
              tagName: 'canvas',
              attributes: {},
              childNodes: [],
            },
          },
        ],
      },
    };
    const events = [...inlineStyleEvents(), mutationEvent] as eventWithTime[];
    const result = pruneBranches(events, { keep: [99] });
    assertSnapshot(result);
  });
  it('should remove branches based on child nodes that came in after fullsnapshot', () => {
    const mutationEvent = {
      type: EventType.IncrementalSnapshot,
      data: {
        source: 0,
        texts: [],
        attributes: [],
        removes: [],
        adds: [
          {
            parentId: 14,
            nextId: null,
            node: {
              id: 98,
              type: NodeType.Element,
              tagName: 'main',
              attributes: {},
              childNodes: [
                {
                  id: 99,
                  type: NodeType.Element,
                  tagName: 'canvas',
                  attributes: {},
                  childNodes: [],
                },
              ],
            },
          },
        ],
      },
    };
    const events = [...inlineStyleEvents(), mutationEvent] as eventWithTime[];
    const result = pruneBranches(events, { keep: [99] });
    assertSnapshot(result);
  });

  it('should keep branches where target child nodes was, and gets moved to', () => {
    const mutationEvent = {
      type: EventType.IncrementalSnapshot,
      data: {
        source: 0,
        texts: [],
        attributes: [],
        removes: [
          {
            parentId: 14,
            id: 15,
          },
        ],
        adds: [
          {
            parentId: 5,
            nextId: null,
            node: { type: 3, textContent: '      \n    ', id: 15 },
          },
        ],
      },
    };
    const events = [...inlineStyleEvents(), mutationEvent] as eventWithTime[];
    const result = pruneBranches(events, { keep: [15] });
    assertSnapshot(result);
  });
});
