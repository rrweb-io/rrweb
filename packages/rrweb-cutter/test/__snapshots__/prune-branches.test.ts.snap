// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`prune branches should cut branches that doesn't include id: pruned all but 14 1`] = `
"1 RRDocument
  -1 RRDocumentType
  2 HTML 
    12 BODY 
      14 DIV id=\\"root\\" 
        15 RRText text=\\"      \\\\n    \\"
"
`;

exports[`prune branches should keep branches where target child nodes was, and gets moved to 1`] = `
"[
  {
    \\"type\\": 0,
    \\"data\\": {}
  },
  {
    \\"type\\": 1,
    \\"data\\": {}
  },
  {
    \\"type\\": 4,
    \\"data\\": {
      \\"href\\": \\"about:blank\\",
      \\"width\\": 1512,
      \\"height\\": 395
    }
  },
  {
    \\"type\\": 2,
    \\"data\\": {
      \\"node\\": {
        \\"type\\": 0,
        \\"childNodes\\": [
          {
            \\"type\\": 2,
            \\"tagName\\": \\"html\\",
            \\"attributes\\": {},
            \\"childNodes\\": [
              {
                \\"type\\": 2,
                \\"tagName\\": \\"head\\",
                \\"attributes\\": {},
                \\"childNodes\\": [
                  {
                    \\"type\\": 2,
                    \\"tagName\\": \\"meta\\",
                    \\"attributes\\": {
                      \\"charset\\": \\"utf-8\\"
                    },
                    \\"childNodes\\": [],
                    \\"id\\": 5
                  }
                ],
                \\"id\\": 3
              },
              {
                \\"type\\": 2,
                \\"tagName\\": \\"body\\",
                \\"attributes\\": {},
                \\"childNodes\\": [
                  {
                    \\"type\\": 2,
                    \\"tagName\\": \\"div\\",
                    \\"attributes\\": {
                      \\"id\\": \\"root\\"
                    },
                    \\"childNodes\\": [
                      {
                        \\"type\\": 3,
                        \\"textContent\\": \\"      \\\\n    \\",
                        \\"id\\": 15
                      }
                    ],
                    \\"id\\": 14
                  }
                ],
                \\"id\\": 12
              }
            ],
            \\"id\\": 2
          }
        ],
        \\"compatMode\\": \\"BackCompat\\",
        \\"id\\": 1
      },
      \\"initialOffset\\": {
        \\"left\\": 0,
        \\"top\\": 0
      }
    }
  },
  {
    \\"type\\": 3,
    \\"data\\": {
      \\"source\\": 2,
      \\"type\\": 2,
      \\"id\\": 15
    }
  },
  {
    \\"type\\": 3,
    \\"data\\": {
      \\"source\\": 0,
      \\"texts\\": [],
      \\"attributes\\": [],
      \\"removes\\": [
        {
          \\"parentId\\": 14,
          \\"id\\": 15
        }
      ],
      \\"adds\\": [
        {
          \\"parentId\\": 5,
          \\"nextId\\": null,
          \\"node\\": {
            \\"type\\": 3,
            \\"textContent\\": \\"      \\\\n    \\",
            \\"id\\": 15
          }
        }
      ]
    }
  }
]"
`;

exports[`prune branches should remove branches based on child nodes that came in after fullsnapshot 1`] = `
"[
  {
    \\"type\\": 0,
    \\"data\\": {}
  },
  {
    \\"type\\": 1,
    \\"data\\": {}
  },
  {
    \\"type\\": 4,
    \\"data\\": {
      \\"href\\": \\"about:blank\\",
      \\"width\\": 1512,
      \\"height\\": 395
    }
  },
  {
    \\"type\\": 2,
    \\"data\\": {
      \\"node\\": {
        \\"type\\": 0,
        \\"childNodes\\": [
          {
            \\"type\\": 2,
            \\"tagName\\": \\"html\\",
            \\"attributes\\": {},
            \\"childNodes\\": [
              {
                \\"type\\": 2,
                \\"tagName\\": \\"body\\",
                \\"attributes\\": {},
                \\"childNodes\\": [
                  {
                    \\"type\\": 2,
                    \\"tagName\\": \\"div\\",
                    \\"attributes\\": {
                      \\"id\\": \\"root\\"
                    },
                    \\"childNodes\\": [],
                    \\"id\\": 14
                  }
                ],
                \\"id\\": 12
              }
            ],
            \\"id\\": 2
          }
        ],
        \\"compatMode\\": \\"BackCompat\\",
        \\"id\\": 1
      },
      \\"initialOffset\\": {
        \\"left\\": 0,
        \\"top\\": 0
      }
    }
  },
  {
    \\"type\\": 3,
    \\"data\\": {
      \\"source\\": 0,
      \\"texts\\": [],
      \\"attributes\\": [],
      \\"removes\\": [],
      \\"adds\\": [
        {
          \\"parentId\\": 14,
          \\"nextId\\": null,
          \\"node\\": {
            \\"id\\": 98,
            \\"type\\": 2,
            \\"tagName\\": \\"main\\",
            \\"attributes\\": {},
            \\"childNodes\\": [
              {
                \\"id\\": 99,
                \\"type\\": 2,
                \\"tagName\\": \\"canvas\\",
                \\"attributes\\": {},
                \\"childNodes\\": []
              }
            ]
          }
        }
      ]
    }
  }
]"
`;

exports[`prune branches should remove branches based on nodes that came in after fullsnapshot 1`] = `
"[
  {
    \\"type\\": 0,
    \\"data\\": {}
  },
  {
    \\"type\\": 1,
    \\"data\\": {}
  },
  {
    \\"type\\": 4,
    \\"data\\": {
      \\"href\\": \\"about:blank\\",
      \\"width\\": 1512,
      \\"height\\": 395
    }
  },
  {
    \\"type\\": 2,
    \\"data\\": {
      \\"node\\": {
        \\"type\\": 0,
        \\"childNodes\\": [
          {
            \\"type\\": 2,
            \\"tagName\\": \\"html\\",
            \\"attributes\\": {},
            \\"childNodes\\": [
              {
                \\"type\\": 2,
                \\"tagName\\": \\"body\\",
                \\"attributes\\": {},
                \\"childNodes\\": [
                  {
                    \\"type\\": 2,
                    \\"tagName\\": \\"div\\",
                    \\"attributes\\": {
                      \\"id\\": \\"root\\"
                    },
                    \\"childNodes\\": [],
                    \\"id\\": 14
                  }
                ],
                \\"id\\": 12
              }
            ],
            \\"id\\": 2
          }
        ],
        \\"compatMode\\": \\"BackCompat\\",
        \\"id\\": 1
      },
      \\"initialOffset\\": {
        \\"left\\": 0,
        \\"top\\": 0
      }
    }
  },
  {
    \\"type\\": 3,
    \\"data\\": {
      \\"source\\": 0,
      \\"texts\\": [],
      \\"attributes\\": [],
      \\"removes\\": [],
      \\"adds\\": [
        {
          \\"parentId\\": 14,
          \\"nextId\\": null,
          \\"node\\": {
            \\"id\\": 99,
            \\"type\\": 2,
            \\"tagName\\": \\"canvas\\",
            \\"attributes\\": {},
            \\"childNodes\\": []
          }
        }
      ]
    }
  }
]"
`;

exports[`prune branches should remove mutations that don't include ids 1`] = `
Object {
  "data": Object {
    "adds": Array [
      Object {
        "nextId": null,
        "node": Object {},
        "parentId": 14,
      },
    ],
    "attributes": Array [
      Object {
        "attributes": Object {
          "data-attr": "Kept",
        },
        "id": 14,
      },
    ],
    "removes": Array [
      Object {
        "id": 15,
        "parentId": 14,
      },
    ],
    "source": 0,
    "texts": Array [
      Object {
        "id": 15,
        "value": "Kept",
      },
    ],
  },
  "type": 3,
}
`;
